Added ecs_taskdefinition_facts module

gurumaia opened the pull request:
##### SUMMARY
Added a module called ecs_taskdefinition_facts that fetches facts about an AWS ECS Task Definition.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
ecs_taskdefinition_facts

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 2b09268473) last updated 2017/09/22 09:58:45 (GMT -200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/gustavo/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/gustavo/Documents/foss/fork/ansible/lib/ansible
  executable location = /home/gustavo/Documents/foss/fork/ansible/bin/ansible
  python version = 3.5.2 (default, Aug 18 2017, 17:48:00) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION
I've referenced both @Java1Guy and @kaczynskid as authors because I basically copied their ecs_service_facts module and adapted for task definitions, so there's a bunch of their code in here.

gurumaia committed:
Added ecs_taskdefinition_facts module

gurumaia committed:
Fix for pep rules

gurumaia committed:
Expanding documentation
Now includes all possible return values

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition_facts.py:
```diff
@@ -0,0 +1,346 @@
+
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+
+    if not HAS_BOTO:
```
Thanks for the contribution! Since new AWS modules are required to just use boto3, you can remove this. Your exception handling can also be updated to be for boto3. Each boto3 call should be in a try/except. There's some helpful info here: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md

gurumaia replied:
Hi @s-hertel, I've removed the boto dependency, leaving only boto3. After a bunch of horrible mistakes on the documentation syntax, the build seems to be stable now. Do you want me to squash my commits or should I leave it like that?

gurumaia committed:
Fixed boto dependency

gurumaia committed:
More fixes to documentation

gurumaia committed:
Fix another typo in documentation

willthames commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition_facts.py:
```diff
@@ -0,0 +1,346 @@
+            if not region:
+                module.fail_json(msg="Region must be specified as a parameter, in EC2_REGION or AWS_REGION environment variables or in boto configuration file")
+            self.ecs = boto3_conn(module, conn_type='client', resource='ecs', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+        except boto3.exception.NoAuthHandlerFound as e:
```
This doesn't actually exist. And won't ever be thrown.

The only error you'll get here is botocore.exceptions.ProfileNotFound

gurumaia replied:
Changed the exception caught. The region check seems to be necessary (it failed non-gracefully without it).

willthames commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition_facts.py:
```diff
@@ -0,0 +1,346 @@
+from ansible.module_utils.ec2 import boto3_conn, ec2_argument_spec, get_aws_connection_info
+
+
+class EcsTaskManager:
```
This class isn't really providing any value. Usual practice is to create the client in main and then call a describe_task_definitions method.

gurumaia replied:
Agreed. This is my first module and I started with code from another that had a similar class, so I thought it was a pattern. Glad to be rid of it.

willthames commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition_facts.py:
```diff
@@ -0,0 +1,346 @@
+            description: The ID of the target.
+            returned: when present
+            type: string
+placementConstraints:
```
We typically convert return results to snake case. ECS must be weird, it returns dromedaryCase (most things return full CamelCase). Still, I see no reason why camel_dict_to_snake_dict shouldn't work

gurumaia replied:
Alright, should be fixed now.

willthames commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition_facts.py:
```diff
@@ -0,0 +1,346 @@
+
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+        task_definition=dict(required=True, type='str')
```
Is there a case  to be made for `task_definition` being optional, and if not present, returning task definitions for all tasks.

Furthermore, I think I'd consider extending this module to become ecs_task_facts, and return the task data along with the task definition and then describing one or more tasks as options.

gurumaia replied:
About making `task_definition` optional, I believe there is a case to be made, but I honestly don't need it right now and am too swamped to implement it today. Is there an official place to track enhancement requests? If so, I suggest we register that as a possible enhancement and implement it in the future.

About merging this with an ecs_task_facts, I'm not so sure. ECS already makes the distinction between tasks and task_definitions confusing enough, I believe this being in the same module would generate more confusion.

gurumaia committed:
Converting results to snake case.

gurumaia committed:
Remove EcsTaskManager class, move to main()
Remove unnecessary `except` block

willthames commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition_facts.py:
```diff
@@ -0,0 +1,333 @@
+            module.fail_json(msg="Region must be specified as a parameter, in EC2_REGION or AWS_REGION environment variables or in boto configuration file")
+        ecs = boto3_conn(module, conn_type='client', resource='ecs', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except botocore.exceptions.ProfileNotFound as e:
+        module.fail_json(msg="AWS profile not found - %s" % str(e))
```
`str(e)` pretty much contains "AWS profile not found", but not a blocker

gurumaia committed:
Fix botocore dependency import

willthames commented:
This looks great now @gurumaia, thanks for making those changes, it looks so much nicer!

We can merge this as is, I think, and let other improvements come along later if there's a real need for them.

However, if we did allow `task_definition` as optional and this to return a list of task definitions, it would change the interface. Let's see whether we can get a second opinion on that.

gurumaia commented:
Yes, changing it to be optional would not only change the interface but also be a substantial change in functionality. I believe this should be settled before releasing 2.5, but for now it seems to be working nicely. Let me just fix the issue with the redundant exception message and I'll be done with it.

gurumaia committed:
Change botocore import method
Also make Profile exception message less redundant

gurumaia commented:
Alright, looks good on my end now.

gurumaia commented:
@willthames I believe I have a use case for not converting the output to snake_case, instead keeping it as returned from AWS (dromedaryCase).
I'm building a role where I need to make sure I don't override the specific image being used in the task definition, because it is set outside of ansible. So I'm using the output of `ecs_taskdefinition_facts` to generate the input for `ecs_taskdefinition` and there the input for the `containers` parameter is exactly like the AWS API, which uses dromedaryCase. In my specific case, since I just need the image, I might be able to work around this since `image` doesn't contain any upper cases, but for other parameters, such as `mount_points`, this problem would arise.

We can either change this module to use dromedaryCase or change `ecs_taskdefinition` to take in snake_case, since it is still marked as preview. What are your thoughts?

willthames commented:
@gurumaia, I can understand that, certainly. Looking at `ecs_taskdefinition`, we pretty much pass the container definitions straight through to boto, we probably wouldn't want to convert that (too much scope for error).

As such, I would support the `container_definitions` result remaining dromedaryCase as long as everything else is snaked.

The easiest way to achieve that is likely (approximate code):

```
    task_def = get_task_definition(...)
    container_definition = task_def['containerDefinition']
    result = dict((k, camel_dict_to_snake_dict(v)) for (k, v) in task_def if k != 'containerDefinition')
    result['container_definition'] = container_definition
```

Hope that is a reasonable compromise.

gurumaia commented:
@willthames sounds good, except that containerDefinitions is not the only one where `ecs_taskdefinition` does that. It does the same for `volumes` and I guess it'll do the same for any complex data structure. I'm starting to think I should convert just the first level keys to snake_case and let everything else as is.

gurumaia committed:
Changing case conversion of the results
Now converts only the root level keys
Commented is a version that would not convert only container_definitions

gurumaia commented:
I've implemented both solutions and I think I prefer the second one, where I only convert the root level keys and leave the values untouched. I've committed this change but left the other version commented so you can take a look. Let me know what you think and I'll remove the extra section so we can move forward.

ryansb commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition_facts.py:
```diff
@@ -0,0 +1,350 @@
+
+    # ecs_td = ecs.describe_task_definition(taskDefinition=module.params['task_definition'])['taskDefinition']
+    # ecs_td_container_definitions = ecs_td['containerDefinitions']
+    # del ecs_td['containerDefinitions']
```
I think of the options I prefer the currently-implemented (not commented) version that changes the top-level, but leaves sub-options as they are, so IMO you can delete this commented stuff and it looks good to me.

gurumaia committed:
Changing case conversion for good

gurumaia committed:
Remove trailing whitespace

gurumaia commented:
Alright, now I think I'm finally done. Should I squash my commits or do you guys usually merge as is?

s-hertel commented:
We squash when we merge, so it's not necessary.
